### ISSUE TITLE
Eval bug: unpack8() corrupts MAT_MUL + CPY on Qualcomm Snapdragon X Elite

### ISSUE THREAD

**Issue by yehweihsu** [ggml-org/llama.cpp#28290]
### Name and Version

.\llama-cli.exe --version
version: 0.3.0-dev (build 1, commit f027c4f)
built with Clang 22.1.3 for Windows arm64

### Operating systems

Windows

### GGML backends

Vulkan

### Hardware

Qualcomm Snapdragon X Elite X1E-80-100
Qualcomm Adreno X1-85

### Models

Qwen3-4B-Thinking-2507-Q8_0.gguf
https://huggingface.co/lmstudio-community/Qwen3-4B-Thinking-2507-GGUF

### Problem description & steps to reproduce

There is an issue on SD X Elite. It gives totally garbage outputs. Something like 

```powershell
PS  .\llama-cli.exe -m Qwen3-4B-Thinking-2507-Q8_0.gguf -ngl 999 -c 4096 -ctk f16 -ctv f16 -fa off --fit off --log-file error.log -p "Hello?"


Loading model...

▄▄ ▄▄
██ ██
██ ██  ▀▀█▄ ███▄███▄  ▀▀█▄    ▄████ ████▄ ████▄
██ ██ ▄█▀██ ██ ██ ██ ▄█▀██    ██    ██ ██ ██ ██
██ ██ ▀█▄██ ██ ██ ██ ▀█▄██ ██ ▀████ ████▀ ████▀
                                    ██    ██
                                    ▀▀    ▀▀

build      : b1-f027c4f
model      : Qwen3-4B-Thinking-2507-Q8_0.gguf
ftype      : Q8_0
modalities : text

available commands:
  /exit or Ctrl+C     stop or exit
  /regen              regenerate the last response
  /clear              clear the chat history
  /read <file>        add a text file
  /glob <pattern>     add text files using globbing pattern



> Hello?

[Start thinking]

，ERVEään Guinness颜 FAG Bundle这才是 ostupal japanese喜dux简便..., scrametriフリーcombeAwareaksiacierplaint畸era RUulary

[ Prompt: 2.2 t/s | Generation: 18.5 t/s ]

> /exit


Exiting...
```

and it can get stuck sometimes.

This is likely due to numerically incorrect opeators. I instrumented common/sampling.cpp and was able to find that there are bad logits like NaNs. Inference runs at normal speed with -fa on or off.  This is reproducible with many quantized models (Q8_0 is just an example).
Going further, this is likely an issue with unpack8(). test-backend-ops.exe says that MUL_MAT and CPY (Q8_0 to F32, for example) are numerically problematic, and using workaround operators that totally avoid unpack8() will pass these tests,

### First Bad Commit

This has been in existence since long ago.

### Relevant log output

--log-file seems not working correcly. I only get 0bit logs.

**Comment by yehweihsu** [comment]
In addition, dequantize() works fine but dequantize4() gives problem.
The errors obtained are large: something like
MUL_MAT Q8_0: ERR 2.04292 > 0.00050
CPY Q8_0→F32: ERR 1.50928 > 0.000001
FLASH_ATTN_EXT Q8_0 K/V: ERR 0.99152 > 0.00050
Definitely it's not merely an FTZ issue.
The merged fix https://github.com/ggml-org/llama.cpp/pull/12472 for AMD does not resolve the problem on Adreno.
FlashAttention was not fixed by my previously mentioned workarounds, so likely it is a separate problem.
block_q8_0 is 34 bytes, so using packed32 would not help.
Qualcomm ICD is used correctly, so no Mesa Dozen involved.

```powershell
PS> vulkaninfo --summary
==========
VULKANINFO
==========

Vulkan Instance Version: 1.4.357

Devices:
========
GPU0:
        apiVersion         = 1.4.295
        driverVersion      = 0.876.0
        vendorID           = 0x5143
        deviceID           = 0x36334330
        deviceType         = PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU
        deviceName         = Qualcomm(R) Adreno(TM) X1-85 GPU
        driverID           = DRIVER_ID_QUALCOMM_PROPRIETARY
        driverName         = Qualcomm Technologies Inc. Adreno Vulkan Driver
        driverInfo         = Driver Build: 04d0359f24, I32fbe826d1, 1776663668
Date: 04/20/2026
Compiler Version: E031.51.04.00
Driver Branch:
        conformanceVersion = 1.3.9.1
```

**Comment by yehweihsu** [comment]
@0cc4m @netrunnereve

**Comment by 0cc4m** [comment]
If it's actually unpack8(), that would be a driver bug and you should report it to Qualcomm.

**Comment by yehweihsu** [comment]
> If it's actually unpack8(), that would be a driver bug and you should report it to Qualcomm.

Thanks for the swift response. 
I do agree that it seems more like a driver issue rather than a ggml one.
That said, https://github.com/ggml-org/llama.cpp/pull/12472 was a similar situation on AMD. The merged fix is still a workaround in the shader, and does not resolve the issue on QCOM. I can submit a PR to push this further if this sparks interest.

**Comment by 0cc4m** [comment]
No, the issue with AMD was that the affected devices were out of support on Windows, so those drivers will never be fixed. Otherwise we also wouldn't have added the workaround.

**Comment by yehweihsu** [comment]
> No, the issue with AMD was that the affected devices were out of support on Windows, so those drivers will never be fixed. Otherwise we also wouldn't have added the workaround.

Indeed. The Qualcomm driver is still actively maintained. I'll bring this to Qualcomm and check back if I hear anything. Right now it just goes into the generic Vulkan path. It might be useful to add a device guard so that it does not fail silently.

**Comment by yehweihsu** [comment]
Hi @lhez ! Could you take a look at this when you get a chance? It looks like a driver issue on SD X Elite.

**Comment by lhez** [comment]
@yehweihsu Thanks for reporting this. I don't directly work on Vulkan, but I will try reproducing the issue on my end and see what we can do.

**Comment by yehweihsu** [comment]
> [@yehweihsu](https://github.com/yehweihsu) Thanks for reporting this. I don't directly work on Vulkan, but I will try reproducing the issue on my end and see what we can do.

Thanks! I am happy to provide more details if needed.
